util/extsort/disk_sorter.go

```diff
@@ -0,0 +1,281 @@
+}
+
+// NewWriter implements the ExternalSorter.NewWriter.
+func (d *DiskSorter) NewWriter(_ context.Context) (Writer, error) {
+	if d.sorted {
+		return nil, errors.Trace(ErrSorted)
+	}
+	return &diskSorterWriter{
+		d:   d,
+		buf: make([]byte, d.opts.WriterBufferSize),
+	}, nil
+}
+
+// Sort implements the ExternalSorter.Sort.
+func (d *DiskSorter) Sort(_ context.Context) error {
+	if d.sorted {
+		return nil
+	}
+
+	iter := d.db.NewIter(nil)
+	if !iter.Last() {
+		// No keys or any error occurred.
+		_ = iter.Close()
+		return errors.Trace(iter.Error())
+	}
+	end := slices.Clone(iter.Key())
+	if err := iter.Close(); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Compact treats end as an exclusive bound,
+	// it doesn't matter since correctness is not affected.
+	err := d.db.Compact(nil, end)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	d.sorted = true
+	return nil
+}
+
+// IsSorted implements the ExternalSorter.IsSorted.
+func (d *DiskSorter) IsSorted() bool {
+	return d.sorted
+}
+
+// NewIterator implements the ExternalSorter.NewIterator.
+func (d *DiskSorter) NewIterator(_ context.Context) (Iterator, error) {
+	if !d.sorted {
+		return nil, errors.Trace(ErrNotSorted)
+	}
+	return &diskSorterIterator{iter: d.db.NewIter(nil)}, nil
+}
+
+// Close implements the ExternalSorter.Close.
+func (d *DiskSorter) Close() error {
+	return errors.Trace(d.db.Close())
+}
+
+// CloseAndCleanup implements the ExternalSorter.CloseAndCleanup.
+func (d *DiskSorter) CloseAndCleanup() error {
+	if err := d.Close(); err != nil {
+		return errors.Trace(err)
+	}
+	fs := d.dbOpts.FS
+	err1 := fs.RemoveAll(d.dbDir)
+	err2 := fs.RemoveAll(d.tmpDir)
+	err := goerrors.Join(err1, err2)
+	return errors.Trace(err)
+}
+
+type diskSorterIterator struct{ iter *pebble.Iterator }
+
+func (i *diskSorterIterator) Seek(key []byte) bool { return i.iter.SeekGE(key) }
+func (i *diskSorterIterator) First() bool          { return i.iter.First() }
+func (i *diskSorterIterator) Next() bool           { return i.iter.Next() }
+func (i *diskSorterIterator) Last() bool           { return i.iter.Last() }
+func (i *diskSorterIterator) Valid() bool          { return i.iter.Valid() }
+func (i *diskSorterIterator) Error() error         { return i.iter.Error() }
+func (i *diskSorterIterator) UnsafeKey() []byte    { return i.iter.Key() }
+func (i *diskSorterIterator) UnsafeValue() []byte  { return i.iter.Value() }
+func (i *diskSorterIterator) Close() error         { return i.iter.Close() }
+
+type keyValue struct {
+	key   []byte
+	value []byte
+}
+
+type diskSorterWriter struct {
+	d   *DiskSorter
+	kvs []keyValue
+	buf []byte
+	off int
+}
+
+func (w *diskSorterWriter) Put(key, value []byte) error {
+	if w.off+len(key)+len(value) > len(w.buf) {
+		if err := w.flush(); err != nil {
+			return errors.Trace(err)
+		}
+		// The default buffer is too small, enlarge it to fit the key and value.
+		if w.off+len(key)+len(value) > len(w.buf) {
+			w.buf = make([]byte, len(key)+len(value))
```

Comment:
Do we need to allocate a larger buffer to reduce the invocation to make? Not sure if that's a hot path